node/metered-channel/src/oneshot.rs

```diff
@@ -0,0 +1,353 @@
+	#[error("No response within {}", Duration::as_secs_f64(.0))]
+	HardTimeout(Duration),
+}
+
+/// Oneshot sender, created by [`channel`].
+#[derive(Debug)]
+pub struct OneshotMeteredSender<T> {
+	name: &'static str,
+	inner: oneshot::Sender<(Instant, T)>,
+	shared_meter: Meter,
+}
+
+impl<T> OneshotMeteredSender<T> {
+	/// Send a value.
+	pub fn send(self, t: T) -> Result<(), T> {
+		let Self { inner, name: _, shared_meter: _ } = self;
+		inner.send((Instant::now(), t)).map_err(|(_, t)| t)
+	}
+
+	/// Poll if the thing is already cancelled.
+	pub fn poll_canceled(&mut self, ctx: &mut Context<'_>) -> Poll<()> {
+		self.inner.poll_canceled(ctx)
+	}
+
+	/// Access the cancellation object.
+	pub fn cancellation(&mut self) -> Cancellation<'_, (Instant, T)> {
+		self.inner.cancellation()
+	}
+
+	/// Check the cancellation state.
+	pub fn is_canceled(&self) -> bool {
+		self.inner.is_canceled()
+	}
+
+	/// Verify if the `receiver` is connected to the `sender` [`Self`].
+	pub fn is_connected_to(&self, receiver: &OneshotMeteredReceiver<T>) -> bool {
+		self.inner.is_connected_to(&receiver.inner)
+	}
+}
+
+/// Oneshot receiver, created by [`channel`].
+#[derive(Debug)]
+pub struct OneshotMeteredReceiver<T> {
+	name: &'static str,
+	inner: oneshot::Receiver<(Instant, T)>,
+	/// Soft timeout, on expire a warning is printed.
+	soft_timeout_fut: Option<Fuse<Delay>>,
+	soft_timeout: Duration,
+	/// Hard timeout, terminating the sender.
+	hard_timeout_fut: Option<Delay>,
+	hard_timeout: Duration,
+	/// The first time the receiver was polled.
+	first_poll_timestamp: Option<Instant>,
+	shared_meter: Meter,
+}
+
+impl<T> OneshotMeteredReceiver<T> {
+	pub fn close(&mut self) {
+		self.inner.close()
+	}
+
+	pub fn try_recv(&mut self) -> Result<Option<T>, Error> {
+		match self.inner.try_recv() {
+			Ok(Some((_when, val))) => Ok(Some(val)),
+			Err(e) => Err(Error::from(e)),
+			Ok(None) => Ok(None),
+		}
+	}
+}
+
+impl<T> OneshotMeteredReceiver<T> {
+	pub fn meter(&self) -> Meter {
+		self.shared_meter.clone()
+	}
+
+	fn update_meter(&mut self, sent_at_timestamp: &Instant, reason: Reason) {
+		let delta: Duration = sent_at_timestamp.elapsed();
+		let val = u64::try_from(delta.as_millis()).unwrap_or(u64::MAX);
+		self.shared_meter
+			.inner
+			.first_poll_till_end_in_millis
+			.store(val, Ordering::Release);
+
+		self.shared_meter.inner.reason.store(reason as u8, Ordering::Release);
+	}
+}
+
+impl<T> FusedFuture for OneshotMeteredReceiver<T> {
+	fn is_terminated(&self) -> bool {
+		self.inner.is_terminated()
+	}
+}
+
+impl<T> Future for OneshotMeteredReceiver<T> {
+	type Output = Result<T, Error>;
+
+	fn poll(mut self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Result<T, Error>> {
+		let first_poll_timestamp =
+			self.first_poll_timestamp.get_or_insert_with(|| Instant::now()).clone();
+
+		let soft_timeout = self.soft_timeout.clone();
+		let soft_timeout = self
+			.soft_timeout_fut
+			.get_or_insert_with(move || Delay::new(soft_timeout).fuse());
```

Comment:
Not sure we need that "on first poll" semantics. Shouldn't it be good enough to have the timeouts since construction? Would reduce overhead quite a bit and I can't think of any realistic use case where this would not suffice.

Reply:
wait - I just saw, that there has been quite some discussion already - let me catch up on that first.

Reply:
Ok, got the argument about first poll/on construction and you could be right - sending the oneshot could actually be blocking on a bounded mpsc. Still, this looks like it should be two types to me. E.g. a `PendingReceiver` which contains the durations and you have a `PendingReceiver::claim(self) -> Receiver` to actually get the receiver for awaiting it:

```rust
let (tx, pending_rx) = metered_oneshot::channel(......);
some_mpsc.send(ImportantRequest(tx)).await;
pending_rx.claim().await
```

 The `Receiver` will be a `PendingReceiver`, but with `Duration`s replaced by `Delay`s.  I am still not getting the meter - let's maybe jump on a call tomorrow on this one.